pennylane/spin/spin_hamiltonian.py

```diff
@@ -93,4 +94,191 @@ def transverse_ising(
         hamiltonian += -h * X(vertex)
 
-    return hamiltonian
+    return hamiltonian.simplify()
+
+
+def heisenberg(lattice, n_cells, coupling=None, boundary_condition=False, neighbour_order=1):
+    r"""Generates the Heisenberg model on a lattice.
+
+    The Hamiltonian is represented as:
+
+    .. math::
+
+         \hat{H} = J\sum_{<i,j>}(\sigma_i^x\sigma_j^x + \sigma_i^y\sigma_j^y + \sigma_i^z\sigma_j^z)
+
+    where ``J`` is the coupling constant defined for the Hamiltonian, and ``i,j`` represent the indices for neighbouring spins.
+
+    Args:
+       lattice (str): Shape of the lattice. Input Values can be ``'chain'``, ``'square'``, ``'rectangle'``,
+                   ``'honeycomb'``, ``'triangle'``, or ``'kagome'``.
+       n_cells (List[int]): Number of cells in each direction of the grid.
+       coupling (List[List[float]] or List[math.array[float]]): Coupling between spins, it can be a 2D array
+                    of shape (neighbour_order, 3) or a 3D array of shape 3 * number of spins * number of spins.
+                    Default value is [1.0, 1.0, 1.0].
+       boundary_condition (bool or list[bool]): Defines boundary conditions for different lattice axes,
+           default is ``False`` indicating open boundary condition.
+       neighbour_order (int): Specifies the interaction level for neighbors within the lattice.
+                    Default is 1, indicating nearest neighbours.
+
+
+    Returns:
+       pennylane.LinearCombination: Hamiltonian for the heisenberg model.
+
+    **Example**
+
+    >>> n_cells = [2,2]
+    >>> j = [[0.5, 0.5, 0.5]]
+    >>> spin_ham = qml.spin.heisenberg("Square", n_cells, coupling=j)
+    >>> spin_ham
+    0.5 * (X(0) @ X(1)) + 0.5 * (Y(0) @ Y(1)) + 0.5 * (Z(0) @ Z(1)) + 0.5 * (X(0) @ X(2)) +
+    0.5 * (Y(0) @ Y(2)) + 0.5 * (Z(0) @ Z(2)) + 0.5 * (X(1) @ X(3)) + 0.5 * (Y(1) @ Y(3)) +
+    0.5 * (Z(1) @ Z(3)) + 0.5 * (X(2) @ X(3)) + 0.5 * (Y(2) @ Y(3)) + 0.5 * (Z(2) @ Z(3))
+    """
+
+    lattice = _generate_lattice(lattice, n_cells, boundary_condition, neighbour_order)
+
+    if coupling is None:
+        coupling = [[1.0, 1.0, 1.0]]
+    coupling = math.asarray(coupling)
+    if coupling.ndim == 1:
+        coupling = math.asarray([coupling])
+
+    if coupling.shape not in [(neighbour_order, 3), (3, lattice.n_sites, lattice.n_sites)]:
+        raise ValueError(
+            f"The coupling parameter shape should be equal to ({neighbour_order},3) or (3,{lattice.n_sites},{lattice.n_sites})"
+        )
+
+    hamiltonian = 0.0 * qml.I(0)
+    if coupling.shape == (neighbour_order, 3):
+        for edge in lattice.edges:
+            i, j, order = edge[0], edge[1], edge[2]
+            hamiltonian += (
+                coupling[order][0] * (X(i) @ X(j))
+                + coupling[order][1] * (Y(i) @ Y(j))
+                + coupling[order][2] * (Z(i) @ Z(j))
+            )
+    else:
+        for edge in lattice.edges:
+            i, j = edge[0], edge[1]
```

Comment:
```suggestion
            i, j = edge
```

Reply:
This won't work, but I see what you are trying to do. I can use slicing here.